In [49]:
import numpy as np
import pandas as pd

In [50]:
data= pd.read_csv(r'G:\Python\diabetes.csv')

In [51]:
data.head()  

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

 Outcome = 1, diabetic and outcome=0, non-diabetic

In [52]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


In [53]:
data.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

#### example : in Glucose 25% means 25% of value are less than 99
    50% of Blood pressure value are less than 72

In [54]:
data.shape  # no. of rows and columns ; the data is taken from 768 people with 9 features

(768, 9)

In [55]:
data['Outcome'].value_counts()

0    500
1    268
Name: Outcome, dtype: int64

In the above case Outcome =0 i.e., 500 people are non-diabetic and Outcome=1 i.e., 268 people are diabetic

In [56]:
data.groupby('Outcome').mean()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
Outcome                                                                      
0           3.298000  109.980000      68.184000      19.664000   68.792000   
1           4.865672  141.257463      70.824627      22.164179  100.335821   

               BMI  DiabetesPedigreeFunction        Age  
Outcome                                                  
0        30.304200                  0.429734  31.190000  
1        35.142537                  0.550500  37.067164

From above observation, People with more age are diabetic, mean age is 37.067

### separating data and labels

In [57]:
x = data.drop(columns='Outcome', axis=1)
y = data['Outcome']

In [58]:
x.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  
0                     0.627   50  
1                     0.351   31  
2                     0.672   32  
3                     0.167   21  
4                     2.288   33

In [59]:
y

0      1
1      0
2      1
3      0
4      1
      ..
763    0
764    0
765    0
766    1
767    0
Name: Outcome, Length: 768, dtype: int64

### Standardization of data

In [60]:
from sklearn.preprocessing import StandardScaler

In [61]:
scaler = StandardScaler()

In [62]:
scaler.fit(x)

StandardScaler()

In [63]:
std_data = scaler.transform(x)

In [64]:
std_data

array([[ 0.63994726,  0.84832379,  0.14964075, ...,  0.20401277,
         0.46849198,  1.4259954 ],
       [-0.84488505, -1.12339636, -0.16054575, ..., -0.68442195,
        -0.36506078, -0.19067191],
       [ 1.23388019,  1.94372388, -0.26394125, ..., -1.10325546,
         0.60439732, -0.10558415],
       ...,
       [ 0.3429808 ,  0.00330087,  0.14964075, ..., -0.73518964,
        -0.68519336, -0.27575966],
       [-0.84488505,  0.1597866 , -0.47073225, ..., -0.24020459,
        -0.37110101,  1.17073215],
       [-0.84488505, -0.8730192 ,  0.04624525, ..., -0.20212881,
        -0.47378505, -0.87137393]])

In [65]:
x = std_data
y = data['Outcome']

### Split the data

In [66]:
from sklearn.model_selection import train_test_split

In [67]:
x_train, x_test, y_train, y_test = train_test_split(x, y , test_size = 0.2, stratify = y, random_state=2)

In [68]:
print(x.shape, x_train.shape, x_test.shape)  # it is good practice to do standardization after training the model

(768, 8) (614, 8) (154, 8)


### Training the model

In [69]:
from sklearn import svm

In [70]:
classifier = svm.SVC(kernel='linear')  # to load svm model into classifier

Training SVM classifier

In [71]:
classifier.fit(x_train, y_train)

SVC(kernel='linear')

### Evaluate model, to check how many times the model is predicting correctly

In [72]:
from sklearn.metrics import accuracy_score

In [73]:
x_train_pred = classifier.predict(x_train)
training_data_accuracy = accuracy_score(x_train_pred, y_train)

In [74]:
training_data_accuracy  # Accuracy score on train data

0.7866449511400652

Accuracy score is 78.66 %, means this model is predicting almost 79 times the correct predictions out of 100 for trained data

In [75]:
x_test_pred = classifier.predict(x_test)
testing_data_accuracy = accuracy_score(x_test_pred, y_test)

In [76]:
testing_data_accuracy  # Accuracy score on test data

0.7727272727272727

Accuracy score is 77.27 %, means this model is predicting almost 77 times the correct predictions out of 100 for test data

This is called overfitting , because training data accuracy is more than test data accuracy

### Make a predictive system that can predict whether a person has diabetes or not by considering the data given.

In [90]:
input_data = (5,117,92,0,0,34.1,0.337,38)

# change input data to nupy array

input_data_as_numpy_array = np.asarray(input_data)
 
# reshape the array, if we donot reshape model expects 768 rows, to avoid this and for giving only one of patient data

input_data_reshaped = input_data_as_numpy_array.reshape(1,-1)

# standardize data before giving it to the model,
# because while training the data we have standardized the data and then trained the model.

std_data = scaler.transform(input_data_reshaped)
print(std_data)

prediction = classifier.predict(std_data)
print(prediction)

[[ 0.3429808  -0.12188771  1.18359575 -1.28821221 -0.69289057  0.26747239
  -0.40734244  0.40494237]]
[0]


In [91]:
if prediction[0] == 0 :
    print("The person is non-diabetic")
else:
    print("The person is diabetic")

The person is non-diabetic
